text/todo-generic-payload.md

```diff
@@ -0,0 +1,368 @@
+
+- Power consumption monitoring
+- Temperature sensing
+- Mass/inertia properties
+- 64-bit value support
+
+## Alternatives
+
+### Alternative 1: Extend Existing Payload Commands
+
+Build upon the current [MAVLink payload services](https://mavlink.io/en/services/payload.html) by adding more generic commands and standardizing implementation.
+
+**Analysis**: The existing `MAV_CMD_DO_SET_ACTUATOR` and `MAV_CMD_DO_SET_SERVO` commands could theoretically be extended with additional parameters for type information and metadata.
+
+**Key Limitations**:
+
+- Commands are stateless and don't provide discovery mechanisms
+- Inconsistent implementation across flight stacks ("only implemented on PX4")
+- Limited parameter space in command messages prevents rich metadata
+- No built-in telemetry streaming capabilities
+- Would require significant breaking changes to existing command semantics
+
+### Alternative 2: Parameter-Only Approach
+
+Use MAVLink's parameter system (`PARAM_*` messages) to expose payload functions as parameters.
+
+**Analysis**: Parameters provide get/set functionality and some metadata (min/max, type). Could potentially expose payload functions as settable parameters.
+
+**Key Limitations**:
+
+- Parameters are designed for configuration, not real-time control
+- No standardized discovery of which parameters belong to payloads
+- Limited data types (mostly float32)
+- No support for momentary operations or complex control modes
+- Parameter namespace pollution with payload-specific entries
+- No built-in telemetry streaming for sensor data
+
+### Alternative 3: Extend Component Information Service
+
+Build upon the [MAVLink Component Information Service](https://mavlink.io/en/services/component_information.html) to include payload metadata and add control messages.
+
+**Analysis**: The component information service provides sophisticated metadata discovery through JSON files and supports actuator configuration. This could potentially be extended to describe payload capabilities.
+
+**Key Limitations**:
+
+- Currently "work in progress" with limited adoption
+- Designed for static configuration metadata, not dynamic control
+- Requires external file hosting or on-device storage for metadata
+- JSON parsing overhead for simple operations
+- No real-time control or telemetry streaming mechanisms
+- Metadata assumed invariant after boot, limiting dynamic payloads
+
+## Unresolved Questions
+
+1. **Message ID Range**: Should we request official MAVLink message IDs or remain in vendor-specific range?
+2. **Backward Compatibility**: How should payloads indicate protocol version support?
+3. **Performance Limits**: What are reasonable limits for number of functions/telemetry channels per payload?
+4. **Rate Control**: How should high-rate telemetry interact with existing MAVLink rate limiting mechanisms?
+5. **Error Recovery**: Should we define standard error recovery procedures for failed commands?
+
```

Comment:
6. Support for FC connected payloads. 
   The camera and gripper (and soon to be winch) can also be attached to an FC. This means the payload can't have a unique component ID, and hence can't be addressed directly. We solved this for camera by having a camera id in the setter commands that can be set in missions to allow addressing of separate cameras in missions (where you don't have control over the the target) and also emitting this in the identity message. The first 6 ids are reserved for FC-attached payloads.

Reply:
Yes, the FC would have to implement code to process incoming messages and broadcast that it supports generic payloads of the given types. Typically it would also process the id in missions and direct the command to the hardware for attached payloads, or reemit the mission item as a command with the id value used as the component id.

An FC wouldn't have to support this - the only restriction this process adds is that generic payloads could not have a component id in the range reserved for those attached to an FC (that's arbitrary - for camera we used 1-6.